Forked typeorm (#6120)
global-121/121-platform: services/121-service/README.md

See instructions to get started in the main [`README`](../../README.md#getting-started).

## Database

> This service uses as [fork](https://github.com/global-121/typeorm) of [TypeORM](https://typeorm.io/) with a PostgreSQL database.

Reviewer: ```suggestion
This service uses a [fork](https://github.com/global-121/typeorm) of [TypeORM](https://typeorm.io/) with a PostgreSQL database.
```

- The 121-service will respond with a (httpOnly-)Cookie containing the users's details and permissions, the cookie will be used automatically on subsequent requests.
- This will give access to each API-endpoint for which a `Permission` is specified and a matching `Permission` is present in the users' token/cookie.

## API

- Access the Swagger UI via: <http://localhost:3000/docs/>
- A graph will be generated when run in 'development' mode at [`module-dependencies.md`](./module-dependencies.md).
  It can be viewed with <https://mermaid.live/> or the VSCode-extension: [Markdown Preview Mermaid Support](https://marketplace.visualstudio.com/items?itemName=bierner.markdown-mermaid)

### Updating/adding Node.js dependencies

Make sure to update any dependencies from _within_ the Docker-container, with:

    docker compose exec 121-service  npm install --save <package-name>

> The only expection to this is TypeORM. To update TypeORM:

Reviewer: ```suggestion
The only exception to this is TypeORM. To update TypeORM:
```

- This will give access to each API-endpoint for which a `Permission` is specified and a matching `Permission` is present in the users' token/cookie.

## API

- Access the Swagger UI via: <http://localhost:3000/docs/>
- A graph will be generated when run in 'development' mode at [`module-dependencies.md`](./module-dependencies.md).
  It can be viewed with <https://mermaid.live/> or the VSCode-extension: [Markdown Preview Mermaid Support](https://marketplace.visualstudio.com/items?itemName=bierner.markdown-mermaid)

### Updating/adding Node.js dependencies

Make sure to update any dependencies from _within_ the Docker-container, with:

    docker compose exec 121-service  npm install --save <package-name>

The only exception to this is TypeORM. To update TypeORM:

- Go to the forked repo and create a new version as described in the [README](https://github.com/global-121/typeorm/)
- Change the version number of typeorm `"typeorm": "npm:@global121/typeorm@<version-number>",` in `services/121-service/package.json` according to the new release.

>   - We cannot use `"@global121/typeorm": "<version-number>",` in the `package.json` because the typeorm package is also a dependency in other packages. This configuration overwrites that.

Reviewer: ```suggestion
  - We cannot use `"@global121/typeorm": "<version-number>",` in the `package.json` because the typeorm package is also a dependency in other packages. This configuration "tricks" npm into treating our fork as if it were the original `typeorm` so that, anywhere in our codebase (including in the `node_modules`), `import ... from 'typeorm'` will use our fork instead of the original `typeorm`
```

Reviewer 2: Aha. that makes a lot of sense. :D

Or we'd need to update all of those import-statements... and back again when TypoORM is updated with the features we'd want...
